vdom/core.py

```diff
@@ -56,157 +52,95 @@ def to_json(el, schema=None):
 
     return json_el
 
-
 class VDOM(object):
     """A basic virtual DOM class which allows you to write literal VDOM spec
 
-    >>> VDOM({ 'tagName': 'h1', 'children': 'Hey', 'attributes': {}})
-
-    It's probably better to use `vdom.h` or the helper functions:
-
-    >>> from vdom import h
-    >>> h('h1', 'Hey')
-    <h1 />
+    >>> VDOM(tag_name='h1', children='Hey', attributes: {}})
 
     >>> from vdom.helpers import h1
     >>> h1('Hey')
-
     """
-    _schema = VDOM_SCHEMA
-    _obj = None
+    # This class should only have these 4 attributes
+    __slots__ = ['tag_name', 'attributes', 'children', 'key']
 
-    def __init__(self, obj, schema=None):
-        # we need to assign self.schema first,
-        # because it is used to validate the object
+    def __init__(self, tag_name, attributes=None, children=None, key=None, schema=None):
         if schema is not None:
-            self.schema = schema
-        self.obj = obj
+            warnings.warn('Passing schema param to VDOM constructor is deprecated')
+        if isinstance(tag_name, dict) or isinstance(tag_name, list):
+            # Backwards compatible interface
+            warnings.warn('Passing dict to VDOM constructor is deprecated')
+            value = tag_name
+            vdom_obj = VDOM.from_dict(value)
+            tag_name = vdom_obj.tag_name
+            attributes = vdom_obj.attributes
+            children = vdom_obj.children
+            key = vdom_obj.key
+        self.tag_name = tag_name
+        self.attributes = attributes if attributes else {}
+        self.children = children if children else []
+        self.key = key
+
+    def to_dict(self):
+        vdom_dict = {
+            'tagName': self.tag_name,
+            'attributes': self.attributes
+        }
+        if self.key:
+            vdom_dict['key'] = self.key
+        vdom_dict['children'] = [c.to_dict() if isinstance(c, VDOM) else c for c in self.children]
+        return vdom_dict
 
-    def _repr_mimebundle_(self, include, exclude, **kwargs):
-        return {'application/vdom.v1+json': self.json_contents}
+    def to_json(self):
```

Comment:
If someone were to modify the contents of the VDOM object after initialising it, would this perform validation?

Reply:
It would not, but IMO VDOM objects should be immutable. That avoids this and many other problems. I was going to make another PR explicitly making these immutable.

Reply:
Are there good data structures in Python for immutability with structural sharing underneath for performance?